docs/man/man3/DNS.3
.TH "DNS" 3 "19 Dec 2005" "Version 1.0Betareleases" "InspIRCd" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DNS \- The DNS class allows fast nonblocking resolution of hostnames and ip addresses.  

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <dns.h>\fP
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDNS\fP ()"
.br
.RI "\fIThe default constructor uses dns addresses read from /etc/resolv.conf. \fP"
.ti -1c
.RI "\fBDNS\fP (\fBstd::string\fP dnsserver)"
.br
.RI "\fIThis constructor accepts a dns server address. \fP"
.ti -1c
.RI "\fB~DNS\fP ()"
.br
.RI "\fIThe destructor frees all used structures. \fP"
.ti -1c
.RI "bool \fBReverseLookup\fP (\fBstd::string\fP ip)"
.br
.RI "\fIThis method will start the reverse lookup of an ip given in dotted decimal format, e.g. \fP"
.ti -1c
.RI "bool \fBForwardLookup\fP (\fBstd::string\fP host)"
.br
.RI "\fIThis method will start the forward lookup of a hostname, e.g. \fP"
.ti -1c
.RI "bool \fBHasResult\fP ()"
.br
.RI "\fIThis method will return true when the lookup is completed. \fP"
.ti -1c
.RI "bool \fBHasResult\fP (int fd)"
.br
.RI "\fIThis method will return true if the lookup's fd matches the one provided. \fP"
.ti -1c
.RI "\fBstd::string\fP \fBGetResult\fP ()"
.br
.RI "\fIThis method returns the result of your query as a string, depending upon wether you called \fBDNS::ReverseLookup()\fP or \fBDNS::ForwardLookup\fP. \fP"
.ti -1c
.RI "\fBstd::string\fP \fBGetResultIP\fP ()"
.br
.ti -1c
.RI "int \fBGetFD\fP ()"
.br
.RI "\fIThis method returns the file handle used by the dns query socket or zero if the query is invalid for some reason, e.g. \fP"
.ti -1c
.RI "void \fBSetNS\fP (\fBstd::string\fP dnsserver)"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBdns_init\fP ()"
.br
.ti -1c
.RI "void \fBdns_init_2\fP (const char *dnsserver)"
.br
.ti -1c
.RI "in_addr * \fBdns_aton4\fP (const char *const ipstring)"
.br
.ti -1c
.RI "char * \fBdns_ntoa4\fP (const in_addr *const ip)"
.br
.ti -1c
.RI "int \fBdns_getip4\fP (const char *const name)"
.br
.ti -1c
.RI "int \fBdns_getip4list\fP (const char *const name)"
.br
.ti -1c
.RI "int \fBdns_getname4\fP (const in_addr *const ip)"
.br
.ti -1c
.RI "char * \fBdns_getresult\fP (const int fd)"
.br
.ti -1c
.RI "in_addr * \fBdns_aton4_s\fP (const char *const ipstring, in_addr *const ip)"
.br
.ti -1c
.RI "char * \fBdns_ntoa4_s\fP (const in_addr *const ip, char *const \fBresult\fP)"
.br
.ti -1c
.RI "char * \fBdns_getresult_s\fP (const int fd, char *const \fBresult\fP)"
.br
.ti -1c
.RI "in_addr * \fBdns_aton4_r\fP (const char *const ipstring)"
.br
.ti -1c
.RI "char * \fBdns_ntoa4_r\fP (const in_addr *const ip)"
.br
.ti -1c
.RI "char * \fBdns_getresult_r\fP (const int fd)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "in_addr * \fBbinip\fP"
.br
.ti -1c
.RI "char * \fBresult\fP"
.br
.ti -1c
.RI "char \fBlocalbuf\fP [1024]"
.br
.ti -1c
.RI "int \fBt\fP"
.br
.ti -1c
.RI "int \fBmyfd\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
The DNS class allows fast nonblocking resolution of hostnames and ip addresses. 

It is based heavily upon firedns by Ian Gulliver.
.PP
Definition at line 35 of file dns.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "DNS::DNS ()"
.PP
The default constructor uses dns addresses read from /etc/resolv.conf. 
.PP
Please note that it will re-read /etc/resolv.conf for each copy of the class you instantiate, causing disk access and slow lookups if you create a lot of them. Consider passing the constructor a server address as a parameter instead.
.SS "DNS::DNS (\fBstd::string\fP dnsserver)"
.PP
This constructor accepts a dns server address. 
.PP
The address must be in dotted decimal form, e.g. 1.2.3.4.
.SS "DNS::~DNS ()"
.PP
The destructor frees all used structures. 
.PP
.SH "Member Function Documentation"
.PP 
.SS "in_addr* DNS::dns_aton4 (const char *const  ipstring)\fC [private]\fP"
.PP
.SS "in_addr* DNS::dns_aton4_r (const char *const  ipstring)\fC [private]\fP"
.PP
.SS "in_addr* DNS::dns_aton4_s (const char *const  ipstring, in_addr *const  ip)\fC [private]\fP"
.PP
.SS "int DNS::dns_getip4 (const char *const  name)\fC [private]\fP"
.PP
.SS "int DNS::dns_getip4list (const char *const  name)\fC [private]\fP"
.PP
.SS "int DNS::dns_getname4 (const in_addr *const  ip)\fC [private]\fP"
.PP
.SS "char* DNS::dns_getresult (const int fd)\fC [private]\fP"
.PP
.SS "char* DNS::dns_getresult_r (const int fd)\fC [private]\fP"
.PP
.SS "char* DNS::dns_getresult_s (const int fd, char *const  result)\fC [private]\fP"
.PP
.SS "void DNS::dns_init ()\fC [private]\fP"
.PP
.SS "void DNS::dns_init_2 (const char * dnsserver)\fC [private]\fP"
.PP
.SS "char* DNS::dns_ntoa4 (const in_addr *const  ip)\fC [private]\fP"
.PP
.SS "char* DNS::dns_ntoa4_r (const in_addr *const  ip)\fC [private]\fP"
.PP
.SS "char* DNS::dns_ntoa4_s (const in_addr *const  ip, char *const  result)\fC [private]\fP"
.PP
.SS "bool DNS::ForwardLookup (\fBstd::string\fP host)"
.PP
This method will start the forward lookup of a hostname, e.g. 
.PP
www.inspircd.org, and returns true if the lookup was successfully initiated.
.SS "int DNS::GetFD ()"
.PP
This method returns the file handle used by the dns query socket or zero if the query is invalid for some reason, e.g. 
.PP
the dns server not responding.
.SS "\fBstd::string\fP DNS::GetResult ()"
.PP
This method returns the result of your query as a string, depending upon wether you called \fBDNS::ReverseLookup()\fP or \fBDNS::ForwardLookup\fP. 
.PP
.SS "\fBstd::string\fP DNS::GetResultIP ()"
.PP
.SS "bool DNS::HasResult (int fd)"
.PP
This method will return true if the lookup's fd matches the one provided. 
.PP
.SS "bool DNS::HasResult ()"
.PP
This method will return true when the lookup is completed. 
.PP
It uses poll internally to determine the status of the socket.
.SS "bool DNS::ReverseLookup (\fBstd::string\fP ip)"
.PP
This method will start the reverse lookup of an ip given in dotted decimal format, e.g. 
.PP
1.2.3.4, and returns true if the lookup was successfully initiated.
.SS "void DNS::SetNS (\fBstd::string\fP dnsserver)"
.PP
.SH "Member Data Documentation"
.PP 
.SS "in_addr* \fBDNS::binip\fP\fC [private]\fP"
.PP
Definition at line 38 of file dns.h.
.SS "char \fBDNS::localbuf\fP[1024]\fC [private]\fP"
.PP
Definition at line 40 of file dns.h.
.SS "int \fBDNS::myfd\fP\fC [private]\fP"
.PP
Definition at line 43 of file dns.h.
.SS "char* \fBDNS::result\fP\fC [private]\fP"
.PP
Definition at line 39 of file dns.h.
.SS "int \fBDNS::t\fP\fC [private]\fP"
.PP
Definition at line 41 of file dns.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for InspIRCd from the source code.


docs/ChangeLog
1.0 Beta 4

    * Fixed configure bug resulting in osflags being called BEFORE getcache (Making it redundant.)
    * Fixed output of src/modules/Makefile saying it was made by Craig.net tongue.gif
    * Updated for test gcc4 compatibility
    * Added support back for ./inspircd
    * Added a line to automatically chmod ./inspircd to 0744 ('Last Second Save!')
    * configure now supports any future .file.inc includes (Yes, i can see into the future)
    * Updated makefile
    * New Perl ./configure
    * Added -Woverloaded-virtual to catch annoying module typos
    * Fixed to always look for the ircd.log within the bin/ dir
    * Fixed to not spool the error log after ./inspircd debug
    * Sanity checks for fd_ref in addclient
    * Fixed 'server ignores me forever if i throttle the server with connections' bug
    * Debugging now allowed from ./inspircd (./inspircd debug)
    * Added LD_LIBRARY_PATH hacks/kludges to the .inspircd.inc file to make it find libmysql in funny situations
    * Added final documentation for inspircd Beta 4
    * Added important note relating to ip addresses in <link:ipaddr>
    * Added schema for m_sqllog.so
    * Added m_sqloper, allows storage of opers within a mysql database
    * Commented new functions
    * Major *MAJOR* optimizations by double-referencing channels to users (never need to scan the entire user hash again except in very rare circumstances)
    * Added channel 'counter', increases speed of quits, parts, kicks
    * commented some users and connections code
    * added remote @* notice
    * Added remote versioning
    * Fixed dicky buffer (due to cleanup the wrong variable was being used)
    * Services speedups for stupid buffered i/o uplinks
    * Added server input buffers (for systems that decide buffering is fun)
    * Major code tidyup (-W)
    * Fixed some flood checking
    * added noticeall
    * Added a Server:: method to send server notices to channels
    * Added explicit oper tracking in a vector of userrec* pointers, optimizes sending out oper notices
    * Fixed: +s and +p channels wouldn't be synched correctly
    * Fixed: +s and +p never show in whois even when you're a member of them
    * Fixed: +s and +p never show in list even if you're a member of them
    * Added Server::FindDescriptor
    * Added and tested fd crossreference table - improves speed a lot, and fixes one of the /STATS chars which has been broken a while
    * Added OnAddBan and OnDelBan module api calls, and fixed glitch which required them in m_timedbans module
    * Added OnOperCompare function to override strcmp in password check for /oper
    * Added module message passing architecture
    * Added a few anti-flood features
    * Updated buffering, faster and more sensible (old code sucks)
    * Added password field
    * Removed some debug output
    * Added /stats T
    * Empty but non-null string passed to is_uline now returns true
    * Added numeric 318 to /WHOIS when nick is not found (end of whois should still be there as its a list response)
    * Added numeric 412 "no text to send"
    * Changed to handle : as a NOP token so that 'assuming' inspircd knows RFC commands wont cause tons of errors
    * Fix to a mode parsing bug (eating wrong version of parameter?)
    * fixed warning of redefinition of RUSAGE_SELF
    * Fix annoying off-by-one error in connection::RecvPacket
    * Added more fault tolerance to server links

src/dns.cpp
/*
dns.cpp - based on the firedns library Copyright (C) 2002 Ian Gulliver

This program is free software; you can redistribute it and/or modify
it under the terms of version 2 of the GNU General Public License as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#define _DNS_C

using namespace std;

#include <string>
#include <stdlib.h>
#include <time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <poll.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "dns.h"
#include "inspircd.h"
#include "helperfuncs.h"
#include "socketengine.h"

extern InspIRCd* ServerInstance;
extern ServerConfig* Config;

#define max(a,b) (a > b ? a : b)
#define DNS_MAX              8                    /* max number of nameservers used */
#define DNS_CONFIG_FBCK     "/etc/resolv.conf"    /* fallback config file */
#define DNS_PORT            53                    /* DNS well known port */
#define DNS_QRY_A            1                    /* name to IP address */
#define DNS_QRY_AAAA        28                    /* name to IP6 address */
#define DNS_QRY_PTR         12                    /* IP address to name */
#define DNS_QRY_MX          15                    /* name to MX */
#define DNS_QRY_TXT         16                    /* name to TXT */
#define DNS_QRY_CNAME       5

#define DNS_ALIGN (sizeof(void *) > sizeof(long) ? sizeof(void *) : sizeof(long))
#define DNS_TRIES 3
#define RESULTSIZE 1024
#define min(a,b) (a < b ? a : b)

static struct in_addr servers4[DNS_MAX]; /* up to DNS_MAX nameservers; populated by dns_init() */
static int i4; /* actual count of nameservers; set by dns_init() */

static int initdone = 0; /* to ensure dns_init() only runs once (on the first call) */
static int wantclose = 0;
static int lastcreate = -1;

struct s_connection { /* open DNS query */
	struct s_connection *next; /* next in list */
	unsigned char id[2];
	unsigned int _class;
	unsigned int type;
	int want_list;
	int fd; /* file descriptor returned from sockets */
};

struct s_rr_middle {
	unsigned int type;
	unsigned int _class;
	unsigned long ttl;
	unsigned int rdlength;
};

#define DNS_POINTER_VALUE 0xc000

static s_connection *connection_head = NULL; /* linked list of open DNS queries; populated by dns_add_query(), decimated by dns_getresult_s() */

struct s_header { /* DNS query header */
	unsigned char id[2];
	unsigned int flags1;
#define FLAGS1_MASK_QR 0x80
#define FLAGS1_MASK_OPCODE 0x78 /* bitshift right 3 */
#define FLAGS1_MASK_AA 0x04
#define FLAGS1_MASK_TC 0x02
#define FLAGS1_MASK_RD 0x01
	unsigned int flags2;
#define FLAGS2_MASK_RA 0x80
#define FLAGS2_MASK_Z  0x70
#define FLAGS2_MASK_RCODE 0x0f
	unsigned int qdcount;
	unsigned int ancount;
	unsigned int nscount;
	unsigned int arcount;
	unsigned char payload[512]; /* DNS question, populated by dns_build_query_payload() */
};

extern time_t TIME;

void *dns_align(void *inp) {
	char *p = (char*)inp;
	int offby = ((char *)p - (char *)0) % DNS_ALIGN;
	if (offby != 0)
		return p + (DNS_ALIGN - offby);
	else
		return p;
}

/*
 * These little hacks are here to avoid alignment and type sizing issues completely by doing manual copies
 */
void dns_fill_rr(s_rr_middle* rr, const unsigned char *input) {
	rr->type = input[0] * 256 + input[1];
	rr->_class = input[2] * 256 + input[3];
	rr->ttl = input[4] * 16777216 + input[5] * 65536 + input[6] * 256 + input[7];
	rr->rdlength = input[8] * 256 + input[9];
}

void dns_fill_header(s_header *header, const unsigned char *input, const int l) {
	header->id[0] = input[0];
	header->id[1] = input[1];
	header->flags1 = input[2];
	header->flags2 = input[3];
	header->qdcount = input[4] * 256 + input[5];
	header->ancount = input[6] * 256 + input[7];
	header->nscount = input[8] * 256 + input[9];
	header->arcount = input[10] * 256 + input[11];
	memcpy(header->payload,&input[12],l);
}

void dns_empty_header(unsigned char *output, const s_header *header, const int l) {
	output[0] = header->id[0];
	output[1] = header->id[1];
	output[2] = header->flags1;
	output[3] = header->flags2;
	output[4] = header->qdcount / 256;
	output[5] = header->qdcount % 256;
	output[6] = header->ancount / 256;
	output[7] = header->ancount % 256;
	output[8] = header->nscount / 256;
	output[9] = header->nscount % 256;
	output[10] = header->arcount / 256;
	output[11] = header->arcount % 256;
	memcpy(&output[12],header->payload,l);
}

void dns_close(int fd) { /* close query */
	log(DEBUG,"DNS: dns_close on fd %d",fd);
	if (fd == lastcreate) {
		wantclose = 1;
		return;
	}
	shutdown(fd,2);
	close(fd);
	return;
}

void DNS::dns_init() { /* on first call only: populates servers4 struct with up to DNS_MAX nameserver IP addresses from /etc/resolv.conf */
	FILE *f;
	int i;
	in_addr addr4;
	char buf[1024];
	if (initdone == 1)
		return;
	i4 = 0;

	initdone = 1;
	srand((unsigned int) TIME);
	memset(servers4,'\0',sizeof(in_addr) * DNS_MAX);
	f = fopen(DNS_CONFIG_FBCK,"r");
	if (f == NULL)
		return;
	while (fgets(buf,1024,f) != NULL) {
		if (strncmp(buf,"nameserver",10) == 0) {
			i = 10;
			while (buf[i] == ' ' || buf[i] == '\t')
				i++;
			if (i4 < DNS_MAX) {
				if (dns_aton4_s(&buf[i],&addr4) != NULL)
					memcpy(&servers4[i4++],&addr4,sizeof(in_addr));
			}
		}
	}
	fclose(f);
}

void DNS::dns_init_2(const char* dnsserver)
{
        in_addr addr4;
        i4 = 0;
        srand((unsigned int) TIME);
        memset(servers4,'\0',sizeof(in_addr) * DNS_MAX);
        if (dns_aton4_s(dnsserver,&addr4) != NULL)
            memcpy(&servers4[i4++],&addr4,sizeof(in_addr));
}


static int dns_send_requests(const s_header *h, const s_connection *s, const int l)
{
	int i;
	sockaddr_in addr4;
	unsigned char payload[sizeof(s_header)];

	dns_empty_header(payload,h,l);


	i = 0;

	/* otherwise send via standard ipv4 boringness */
	memset(&addr4,0,sizeof(addr4));
	memcpy(&addr4.sin_addr,&servers4[i],sizeof(addr4.sin_addr));
	addr4.sin_family = AF_INET;
	addr4.sin_port = htons(DNS_PORT);
	if (sendto(s->fd, payload, l + 12, 0, (sockaddr *) &addr4, sizeof(addr4)) == -1)
	{
		return -1;
	}

	return 0;
}

static s_connection *dns_add_query(s_header *h) { /* build DNS query, add to list */
	s_connection * s;

	s = new s_connection;

	/* set header flags */
	h->id[0] = s->id[0] = rand() % 255; /* verified by dns_getresult_s() */
	h->id[1] = s->id[1] = rand() % 255;
	h->flags1 = 0 | FLAGS1_MASK_RD;
	h->flags2 = 0;
	h->qdcount = 1;
	h->ancount = 0;
	h->nscount = 0;
	h->arcount = 0;

	/* turn off want_list by default */
	s->want_list = 0;

	/* try to create ipv6 or ipv4 socket */
		s->fd = socket(PF_INET, SOCK_DGRAM, 0);
		if (s->fd != -1) {
			if (fcntl(s->fd, F_SETFL, O_NONBLOCK) != 0) {
				shutdown(s->fd,2);
				close(s->fd);
				s->fd = -1;
			}
		}
		if (s->fd != -1) {
			sockaddr_in addr;
			memset(&addr,0,sizeof(addr));
			addr.sin_family = AF_INET;
			addr.sin_port = 0;
			addr.sin_addr.s_addr = INADDR_ANY;
			if (bind(s->fd,(sockaddr *)&addr,sizeof(addr)) != 0) {
				shutdown(s->fd,2);
				close(s->fd);
				s->fd = -1;
			}
		}
		if (s->fd == -1) {
			delete s;
			return NULL;
		}
	/* create new connection object, add to linked list */
	s->next = connection_head;
	connection_head = s;

	if (wantclose == 1) {
		shutdown(lastcreate,2);
		close(lastcreate);
		wantclose = 0;
	}
	lastcreate = s->fd;
	return s;
}

static int dns_build_query_payload(const char * const name, const unsigned short rr, const unsigned short _class, unsigned char * const payload) { 
	short payloadpos;
	const char * tempchr, * tempchr2;
	unsigned short l;

	payloadpos = 0;
	tempchr2 = name;

	/* split name up into labels, create query */
	while ((tempchr = strchr(tempchr2,'.')) != NULL) {
		l = tempchr - tempchr2;
		if (payloadpos + l + 1 > 507)
			return -1;
		payload[payloadpos++] = l;
		memcpy(&payload[payloadpos],tempchr2,l);
		payloadpos += l;
		tempchr2 = &tempchr[1];
	}
	l = strlen(tempchr2);
	if (l) {
		if (payloadpos + l + 2 > 507)
			return -1;
		payload[payloadpos++] = l;
		memcpy(&payload[payloadpos],tempchr2,l);
		payloadpos += l;
		payload[payloadpos++] = '\0';
	}
	if (payloadpos > 508)
		return -1;
	l = htons(rr);
	memcpy(&payload[payloadpos],&l,2);
	l = htons(_class);
	memcpy(&payload[payloadpos + 2],&l,2);
	return payloadpos + 4;
}

in_addr* DNS::dns_aton4(const char * const ipstring) { /* ascii to numeric: convert string to static 4part IP addr struct */
	static in_addr ip;
	return dns_aton4_s(ipstring,&ip);
}

in_addr* DNS::dns_aton4_r(const char *ipstring) { /* ascii to numeric (reentrant): convert string to new 4part IP addr struct */
	in_addr* ip;
	ip = new in_addr;
	if(dns_aton4_s(ipstring,ip) == NULL) {
		delete ip;
		return NULL;
	}
	return ip;
}

in_addr* DNS::dns_aton4_s(const char *ipstring, in_addr *ip) { /* ascii to numeric (buffered): convert string to given 4part IP addr struct */
	inet_aton(ipstring,ip);
	return ip;
}

int DNS::dns_getip4(const char *name) { /* build, add and send A query; retrieve result with dns_getresult() */
	s_header h;
	s_connection *s;
	int l;

	dns_init();
	

	l = dns_build_query_payload(name,DNS_QRY_A,1,(unsigned char *)&h.payload);
	if (l == -1)
		return -1;
	s = dns_add_query(&h);
	if (s == NULL)
		return -1;
	s->_class = 1;
	s->type = DNS_QRY_A;
	if (dns_send_requests(&h,s,l) == -1)
		return -1;

	return s->fd;
}

int DNS::dns_getip4list(const char *name) { /* build, add and send A query; retrieve result with dns_getresult() */
	s_header h;
	s_connection *s;
	int l;

	dns_init();
	

	l = dns_build_query_payload(name,DNS_QRY_A,1,(unsigned char *)&h.payload);
	if (l == -1)
		return -1;
	s = dns_add_query(&h);
	if (s == NULL)
		return -1;
	s->_class = 1;
	s->type = DNS_QRY_A;
	s->want_list = 1;
	if (dns_send_requests(&h,s,l) == -1)
		return -1;

	return s->fd;
}

int DNS::dns_getname4(const in_addr *ip) { /* build, add and send PTR query; retrieve result with dns_getresult() */
	char query[512];
	s_header h;
	s_connection * s;
	unsigned char *c;
	int l;

	c = (unsigned char *)&ip->s_addr;

	sprintf(query,"%d.%d.%d.%d.in-addr.arpa",c[3],c[2],c[1],c[0]);

	l = dns_build_query_payload(query,DNS_QRY_PTR,1,(unsigned char *)&h.payload);
	if (l == -1)
		return -1;
	s = dns_add_query(&h);
	if (s == NULL)
		return -1;
	s->_class = 1;
	s->type = DNS_QRY_PTR;
	if (dns_send_requests(&h,s,l) == -1)
		return -1;

	return s->fd;
}

char* DNS::dns_ntoa4(const in_addr * const ip) { /* numeric to ascii: convert 4part IP addr struct to static string */
	static char r[256];
	return dns_ntoa4_s(ip,r);
}

char* DNS::dns_ntoa4_s(const in_addr *ip, char *r) { /* numeric to ascii (buffered): convert 4part IP addr struct to given string */
	unsigned char *m;
	m = (unsigned char *)&ip->s_addr;
	sprintf(r,"%d.%d.%d.%d",m[0],m[1],m[2],m[3]);
	return r;
}

char* DNS::dns_getresult(const int cfd) { /* retrieve result of DNS query */
	log(DEBUG,"DNS: dns_getresult with cfd=%d",cfd);
	return dns_getresult_s(cfd,this->localbuf);
}

char* DNS::dns_getresult_s(const int cfd, char *res) { /* retrieve result of DNS query (buffered) */
	s_header h;
	s_connection *c, *prev;
	int l,i,q,curanswer,o;
	s_rr_middle rr;
	unsigned char buffer[sizeof(s_header)];
	unsigned short p;

	if (res)
	{
		res[0] = 0;
	}

	prev = NULL;
	c = connection_head;
	while (c != NULL) { /* find query in list of open queries */
		if (c->fd == cfd)
			break;
		prev = c;
		c = c->next;
	}
	if (c == NULL) {
		log(DEBUG,"DNS: got a response for a query we didnt send with fd=%d",cfd);
		return NULL; /* query not found */
	}
	/* query found-- pull from list: */
	if (prev != NULL)
		prev->next = c->next;
	else
		connection_head = c->next;

	l = recv(c->fd,buffer,sizeof(s_header),0);
	dns_close(c->fd);
	if (l < 12) {
		delete c;
		return NULL;
	}
	dns_fill_header(&h,buffer,l - 12);
	if (c->id[0] != h.id[0] || c->id[1] != h.id[1]) {
		log(DEBUG,"DNS: id mismatch on query");
		delete c;
		return NULL; /* ID mismatch */
	}
	if ((h.flags1 & FLAGS1_MASK_QR) == 0) {
		log(DEBUG,"DNS: didnt get a query result");
		delete c;
		return NULL;
	}
	if ((h.flags1 & FLAGS1_MASK_OPCODE) != 0) {
		log(DEBUG,"DNS: got an OPCODE and didnt want one");
		delete c;
		return NULL;
	}
	if ((h.flags2 & FLAGS2_MASK_RCODE) != 0) {
		log(DEBUG,"DNS lookup failed due to SERVFAIL");
		delete c;
		return NULL;
	}
	if (h.ancount < 1)  { /* no sense going on if we don't have any answers */
		log(DEBUG,"DNS: no answers!");
		delete c;
		return NULL;
	}
	/* skip queries */
	i = 0;
	q = 0;
	l -= 12;
	while ((unsigned)q < h.qdcount && i < l) {
		if (h.payload[i] > 63) { /* pointer */
			i += 6; /* skip pointer, _class and type */
			q++;
		} else { /* label */
			if (h.payload[i] == 0) {
				q++;
				i += 5; /* skip nil, _class and type */
			} else
				i += h.payload[i] + 1; /* skip length and label */
		}
	}
	/* &h.payload[i] should now be the start of the first response */
	curanswer = 0;
	while ((unsigned)curanswer < h.ancount) {
		q = 0;
		while (q == 0 && i < l) {
			if (h.payload[i] > 63) { /* pointer */
				i += 2; /* skip pointer */
				q = 1;
			} else { /* label */
				if (h.payload[i] == 0) {
					i++;
					q = 1;
				} else
					i += h.payload[i] + 1; /* skip length and label */
			}
		}
		if (l - i < 10) {
			delete c;
			return NULL;
		}
		dns_fill_rr(&rr,&h.payload[i]);
		i += 10;
		if (rr.type != c->type) {
			curanswer++;
			i += rr.rdlength;
			continue;
		}
		if (rr._class != c->_class) {
			curanswer++;
			i += rr.rdlength;
			continue;
		}
		break;
	}
	if ((unsigned)curanswer == h.ancount)
		return NULL;
	if ((unsigned)i + rr.rdlength > (unsigned)l)
		return NULL;
	if (rr.rdlength > 1023)
		return NULL;

	switch (rr.type) {
		case DNS_QRY_PTR:
			log(DEBUG,"DNS: got a result of type DNS_QRY_PTR");
			o = 0;
			q = 0;
			while (q == 0 && i < l && o + 256 < 1023) {
				if (h.payload[i] > 63) { /* pointer */
					memcpy(&p,&h.payload[i],2);
					i = ntohs(p) - DNS_POINTER_VALUE - 12;
				} else { /* label */
					if (h.payload[i] == 0)
						q = 1;
					else {
						res[o] = '\0';
						if (o != 0)
							res[o++] = '.';
						memcpy(&res[o],&h.payload[i + 1],h.payload[i]);
						o += h.payload[i];
						i += h.payload[i] + 1;
					}
				}
			}
			res[o] = '\0';
			break;
		case DNS_QRY_A:
			log(DEBUG,"DNS: got a result of type DNS_QRY_A");
			if (c->want_list) {
				dns_ip4list *alist = (dns_ip4list *) res; /* we have to trust that this is aligned */
				while ((char *)alist - (char *)res < 700) {
					if (rr.type != DNS_QRY_A)
						break;
					if (rr._class != 1)
						break;
					if (rr.rdlength != 4) {
						delete c;
						return NULL;
					}
					memcpy(&alist->ip,&h.payload[i],4);
					if ((unsigned)++curanswer >= h.ancount)
						break;
					i += rr.rdlength;
					{
						/* skip next name */
						q = 0;
						while (q == 0 && i < l) {
							if (h.payload[i] > 63) { /* pointer */
								i += 2; /* skip pointer */
								q = 1;
							} else { /* label */
								if (h.payload[i] == 0) {
									i++;
									q = 1;
								} else
									i += h.payload[i] + 1; /* skip length and label */
							}
						}
					}
					if (l - i < 10) {
						delete c;
						return NULL;
					}
					dns_fill_rr(&rr,&h.payload[i]);
					i += 10;
					alist->next = (dns_ip4list *) dns_align(((char *) alist) + sizeof(dns_ip4list));
					alist = alist->next;
					alist->next = NULL;
				}
				alist->next = NULL;
				break;
			}
			goto defaultcase;
			break;
		default:
		defaultcase:
			log(DEBUG,"DNS: doing something with result 'default'");
			memcpy(res,&h.payload[i],rr.rdlength);
			res[rr.rdlength] = '\0';
			break;
	}
	delete c;
	return res;
}

DNS::DNS()
{
	dns_init();
	log(DEBUG,"Create blank DNS");
}

DNS::DNS(std::string dnsserver)
{
	dns_init_2(dnsserver.c_str());
	log(DEBUG,"Create DNS");
}

void DNS::SetNS(std::string dnsserver)
{
	dns_init_2(dnsserver.c_str());
	log(DEBUG,"Set NS");
}

DNS::~DNS()
{
}

bool DNS::ReverseLookup(std::string ip)
{
	ServerInstance->stats->statsDns++;
        binip = dns_aton4(ip.c_str());
        if (binip == NULL) {
                return false;
        }

        this->myfd = dns_getname4(binip);
	if (this->myfd == -1)
	{
		return false;
	}
	log(DEBUG,"DNS: ReverseLookup, fd=%d",this->myfd);
#ifndef THREADED_DNS
	ServerInstance->SE->AddFd(this->myfd,true,X_ESTAB_DNS);
#endif
	return true;
}

bool DNS::ForwardLookup(std::string host)
{
	ServerInstance->stats->statsDns++;
	this->myfd = dns_getip4(host.c_str());
	if (this->myfd == -1)
	{
		return false;
	}
	log(DEBUG,"DNS: ForwardLookup, fd=%d",this->myfd);
#ifndef THREADED_DNS
	ServerInstance->SE->AddFd(this->myfd,true,X_ESTAB_DNS);
#endif
	return true;
}

bool DNS::HasResult(int fd)
{
	return (fd == this->myfd);
}

/* Only the multithreaded dns uses this poll() based
 * check now. As its in another thread we dont have
 * to worry about its performance that much.
 */
bool DNS::HasResult()
{
	log(DEBUG,"DNS: HasResult, fd=%d",this->myfd);
	pollfd polls;
	polls.fd = this->myfd;
	polls.events = POLLIN;
	int ret = poll(&polls,1,1);
	log(DEBUG,"DNS: Hasresult returning %d",ret);
	return (ret > 0);
}

int DNS::GetFD()
{
	return this->myfd;
}

std::string DNS::GetResult()
{
	log(DEBUG,"DNS: GetResult()");
        result = dns_getresult(this->myfd);
#ifndef THREADED_DNS
	ServerInstance->SE->DelFd(this->myfd);
#endif
        if (result) {
		ServerInstance->stats->statsDnsGood++;
		dns_close(this->myfd);
		return result;
        } else {
		ServerInstance->stats->statsDnsBad++;
		if (this->myfd != -1)
		{
			dns_close(this->myfd);
		}
		return "";
	}
}

std::string DNS::GetResultIP()
{
	char r[1024];
	log(DEBUG,"DNS: GetResultIP()");
	result = dns_getresult(this->myfd);
	if (this->myfd != -1)
	{
#ifndef THREADED_DNS
		ServerInstance->SE->DelFd(this->myfd);
#endif
		dns_close(this->myfd);
	}
	if (result)
	{
		ServerInstance->stats->statsDnsGood++;
		unsigned char a = (unsigned)result[0];
		unsigned char b = (unsigned)result[1];
		unsigned char c = (unsigned)result[2];
		unsigned char d = (unsigned)result[3];
		snprintf(r,1024,"%u.%u.%u.%u",a,b,c,d);
		return r;
	}
	else
	{
		ServerInstance->stats->statsDnsBad++;
		log(DEBUG,"DANGER WILL ROBINSON! NXDOMAIN for forward lookup, but we got a reverse lookup!");
		return "";
	}
}



#ifdef THREADED_DNS
void* dns_task(void* arg)
{
        userrec* u = (userrec*)arg;
        log(DEBUG,"DNS thread for user %s",u->nick);
        DNS dns1;
        DNS dns2;
        std::string host;
        std::string ip;
        if (dns1.ReverseLookup(u->ip))
        {
                while (!dns1.HasResult())
                {
                        usleep(100);
                }
                host = dns1.GetResult();
                if (host != "")
                {
                        if (dns2.ForwardLookup(host))
                        {
                                while (!dns2.HasResult())
                                {
                                        usleep(100);
                                }
                                ip = dns2.GetResultIP();
                                if (ip == std::string(u->ip))
                                {
                                        if (host.length() < 160)
                                        {
                                                strcpy(u->host,host.c_str());
                                                strcpy(u->dhost,host.c_str());
                                        }
                                }
                        }
                }
        }
        u->dns_done = true;
        return NULL;
}
#endif
